File: src/lib/importer.ts
import { promises as fs } from "node:fs";
import { stat } from "node:fs/promises";
import path from "node:path";

// ---------- types from MIT OCW data.json files ----------
type CourseData = {
	course_title?: string;
	course_description?: string;
	site_uid?: string;
	legacy_uid?: string;
	instructors?: Array<{
		first_name?: string;
		last_name?: string;
		middle_initial?: string;
		salutation?: string;
		title?: string;
	}>;
	department_numbers?: string[];
	learning_resource_types?: string[];
	topics?: string[][];
	mit_learn_topics?: string[] | null;
	primary_course_number?: string;
	extra_course_numbers?: string;
	term?: string;
	year?: string;
	level?: string[];
	image_src?: string | null;
	course_image_metadata?: {
		file?: string;
		image_metadata?: { caption?: string };
	};
};

type ResourceData = {
	title?: string;
	description?: string;
	file?: string | null;
	learning_resource_types?: string[];
	resource_type?: string;
	file_type?: string | null;
	youtube_key?: string;
	captions_file?: string;
	transcript_file?: string;
	thumbnail_file?: string;
	archive_url?: string;
};

type PageData = {
	title?: string;
	content?: string;
	description?: string;
	files?: Array<{
		file?: string;
		title?: string;
		description?: string;
		file_type?: string;
	}>;
};

// Resource enriched with resolved disk-backed URLs
type Resolved = {
	pdfUrl: string | null; // /courses/<rel>/static_resources/<basename>
	transcriptUrl: string | null;
	thumbnailUrl: string | null; // could be youtube CDN (http) or local
};

// ---------- session parsing ----------
type ParsedSession = {
	sessionNumber: string | null;
	title: string;
	assignmentMarkers: string[]; // raw strings: "Assignment 1 out", "Assignment 2 due"
};

const ASSIGNMENT_RE = /Assignment\s+(\d+|\w+)\s+(out|due)/gi;

function parseCalendarContent(raw: string): ParsedSession[] {
	if (!raw) return [];
	const lines = raw
		.replace(/\r/g, "")
		.split("\n")
		.map((l) => l.trim())
		.filter(Boolean);

	while (
		lines.length &&
		/^(SES|#|session|topics|key\s+dates)/i.test(lines[0])
	) {
		lines.shift();
	}

	const sessions: ParsedSession[] = [];
	for (const line of lines) {
		const m = line.match(
			/^([A-Za-z]?\s*\d+\s*[-–~]\s*\d+|[A-Za-z]?\s*\d+)\s+(.*)$/,
		);
		if (!m) continue;
		const sessionNumber = m[1].replace(/\s+/g, "").trim();
		const rest = m[2].trim();

		const assignmentMarkers: string[] = [];
		let am: RegExpExecArray | null;
		ASSIGNMENT_RE.lastIndex = 0;
		while ((am = ASSIGNMENT_RE.exec(rest)) !== null) {
			assignmentMarkers.push(`${am[1]} ${am[2]}`.toLowerCase());
		}

		const title = rest
			.replace(ASSIGNMENT_RE, "")
			.replace(/\s{2,}/g, " ")
			.trim();
		sessions.push({ sessionNumber, title, assignmentMarkers });
	}
	return sessions;
}

// ---------- helpers ----------
async function readJson<T = any>(p: string): Promise<T | null> {
	try {
		const buf = await fs.readFile(p, "utf8");
		return JSON.parse(buf) as T;
	} catch {
		return null;
	}
}

async function listSubdirs(dir: string): Promise<string[]> {
	try {
		const entries = await fs.readdir(dir, { withFileTypes: true });
		return entries
			.filter((e) => e.isDirectory())
			.map((e) => e.name)
			.filter((n) => !n.startsWith("."));
	} catch {
		return [];
	}
}

function sessionNumberFromFolder(name: string): number | null {
	const m = name.match(/^lecture[-_ ]?(\d+)/i);
	if (m) return parseInt(m[1], 10);
	const m2 = name.match(/[_-]?lec(\d+)/i);
	if (m2) return parseInt(m2[1], 10);
	return null;
}

function hwNumberFromFolder(name: string): number | null {
	const m = name.match(/[_-]?hw\s*(\d+)/i);
	return m ? parseInt(m[1], 10) : null;
}

function slugify(s: string): string {
	return s
		.toLowerCase()
		.replace(/[^a-z0-9]+/g, "-")
		.replace(/^-+|-+$/g, "")
		.slice(0, 80);
}

// The data.json `file` field looks like:
//   /courses/1-258j-public-transportation-systems-spring-2017/3385fcada36ca48999bf5637bdad70f5_MIT1_258JS17_lec01.pdf
// The actual file lives at {coursePath}/static_resources/{basename}. Verify on disk.
function basenameOf(fileUrl: string | null | undefined): string | null {
	if (!fileUrl) return null;
	const clean = fileUrl.split("?")[0].split("#")[0];
	const base = path.posix.basename(clean);
	return base || null;
}

async function resolveStatic(
	coursePath: string,
	courseRelPath: string,
	fileUrl: string | null | undefined,
): Promise<string | null> {
	const base = basenameOf(fileUrl);
	if (!base) return null;
	const absDisk = path.join(coursePath, "static_resources", base);
	try {
		const s = await stat(absDisk);
		if (!s.isFile()) return null;
	} catch {
		return null;
	}
	return `/courses/${courseRelPath}/static_resources/${base}`;
}

// ---------- main entry ----------
export type ImportResult = {
	courseSlug: string;
	courseTitle: string;
	sessions: number;
	items: number;
	assignments: number;
	warnings: string[];
};

export type ParsedLecture = {
	sessionNumber: number | null;
	sourceKey: string;
	resource: ResourceData;
	resolved: Resolved;
	folderName: string;
};

export type ParsedAssignment = {
	hwNumber: number | null;
	sourceKey: string;
	resource: ResourceData;
	resolved: Resolved;
	folderName: string;
};

export type ParsedCourse = {
	course: {
		slug: string;
		title: string;
		description: string | null;
		courseNumber: string | null;
		extraCourseNumbers: string | null;
		term: string | null;
		year: string | null;
		level: string | null;
		departmentNumbers: string[] | null;
		instructors: CourseData["instructors"];
		topics: string[][] | null;
		imageUrl: string | null;
		sourcePath: string;
		sourceUid: string | null;
	};
	sessions: ParsedSession[];
	lectures: ParsedLecture[];
	assignments: ParsedAssignment[];
	otherItems: Array<{
		sourceKey: string;
		resource: ResourceData;
		resolved: Resolved;
		folderName: string;
		type: string;
	}>;
	assignmentsPageContent: string | null;
	warnings: string[];
	coursePath: string;
	courseRelPath: string;
	folderName: string;
};

export async function findCourseDirs(root: string): Promise<string[]> {
	// BFS one level deep for nested (mit opencourseware/<course>)
	const subdirs = await listSubdirs(root);
	const courses: string[] = [];
	for (const sub of subdirs) {
		const subPath = path.join(root, sub);
		if (await readJson(path.join(subPath, "data.json"))) {
			courses.push(subPath);
			continue;
		}
		const nested = await listSubdirs(subPath);
		for (const n of nested) {
			const np = path.join(subPath, n);
			if (await readJson(path.join(np, "data.json"))) courses.push(np);
		}
	}
	return courses;
}

export async function parseCourse(
	coursePath: string,
	root: string,
): Promise<ParsedCourse | null> {
	const warnings: string[] = [];
	const data = await readJson<CourseData>(path.join(coursePath, "data.json"));
	if (!data) {
		warnings.push(`No data.json at ${coursePath}`);
		return null;
	}

	const folderName = path.basename(coursePath);
	const courseRelPath = path
		.relative(root, coursePath)
		.split(path.sep)
		.join("/");
	const slug = slugify(
		`${data.primary_course_number ?? folderName}-${data.term ?? ""}-${data.year ?? ""}`,
	);

	// Course image — image_src is "./static_resources/foo.jpg" relative to coursePath
	let imageUrl: string | null = null;
	if (data.image_src && /^https?:\/\//.test(data.image_src)) {
		imageUrl = data.image_src;
	} else if (data.image_src) {
		const rel = data.image_src.replace(/^\.?\//, "");
		const absDisk = path.join(coursePath, rel);
		try {
			const s = await stat(absDisk);
			if (s.isFile())
				imageUrl = `/courses/${courseRelPath}/${rel.split(path.sep).join("/")}`;
		} catch {
			/* leave null */
		}
	}

	const course = {
		slug,
		title: data.course_title ?? folderName,
		description: data.course_description ?? null,
		courseNumber: data.primary_course_number ?? null,
		extraCourseNumbers: data.extra_course_numbers ?? null,
		term: data.term ?? null,
		year: data.year ?? null,
		level: data.level?.[0] ?? null,
		departmentNumbers: data.department_numbers ?? null,
		instructors: data.instructors ?? [],
		topics: data.topics ?? null,
		imageUrl,
		sourcePath: courseRelPath,
		sourceUid: data.site_uid ?? null,
	};

	// Calendar
	const calendarData = await readJson<PageData>(
		path.join(coursePath, "pages", "calendar", "data.json"),
	);
	const sessions: ParsedSession[] = calendarData?.content
		? parseCalendarContent(calendarData.content)
		: [];
	const assignmentsData = await readJson<PageData>(
		path.join(coursePath, "pages", "assignments", "data.json"),
	);

	async function resolve(r: ResourceData): Promise<Resolved> {
		const [pdfUrl, transcriptUrl] = await Promise.all([
			resolveStatic(coursePath, courseRelPath, r.file),
			resolveStatic(coursePath, courseRelPath, r.transcript_file),
		]);
		let thumbnailUrl: string | null = null;
		if (r.thumbnail_file && /^https?:\/\//.test(r.thumbnail_file)) {
			thumbnailUrl = r.thumbnail_file;
		} else if (r.thumbnail_file) {
			thumbnailUrl = await resolveStatic(
				coursePath,
				courseRelPath,
				r.thumbnail_file,
			);
		}
		return { pdfUrl, transcriptUrl, thumbnailUrl };
	}

	// Resources
	const resourcesDir = path.join(coursePath, "resources");
	const resourceFolders = await listSubdirs(resourcesDir);

	const lectures: ParsedLecture[] = [];
	const assignments: ParsedAssignment[] = [];
	const otherItems: ParsedCourse["otherItems"] = [];

	// First pass: collect lec-numbered PDFs separately
	const pdfOnlyByLecN = new Map<
		number,
		{ folder: string; r: ResourceData; resolved: Resolved }
	>();

	for (const folder of resourceFolders) {
		const fp = path.join(resourcesDir, folder);
		const r = await readJson<ResourceData>(path.join(fp, "data.json"));
		if (!r) continue;
		const resolved = await resolve(r);
		const lecN = sessionNumberFromFolder(folder);
		const hwN = hwNumberFromFolder(folder);

		const isPdf =
			r.file_type === "application/pdf" || r.resource_type === "Document";

		if (lecN !== null && isPdf) {
			pdfOnlyByLecN.set(lecN, { folder, r, resolved });
			continue;
		}
		if (hwN !== null && isPdf) {
			assignments.push({
				hwNumber: hwN,
				sourceKey: folder,
				resource: r,
				resolved,
				folderName: folder,
			});
			continue;
		}
		if (lecN !== null) {
			lectures.push({
				sessionNumber: lecN,
				sourceKey: folder,
				resource: r,
				resolved,
				folderName: folder,
			});
		} else if (r.youtube_key) {
			lectures.push({
				sessionNumber: null,
				sourceKey: folder,
				resource: r,
				resolved,
				folderName: folder,
			});
		} else if (hwN !== null) {
			assignments.push({
				hwNumber: hwN,
				sourceKey: folder,
				resource: r,
				resolved,
				folderName: folder,
			});
		} else {
			otherItems.push({
				sourceKey: folder,
				resource: r,
				resolved,
				folderName: folder,
				type: r.resource_type ?? "other",
			});
		}
	}

	// Attach lec PDFs to video lectures
	for (const lec of lectures) {
		if (lec.sessionNumber !== null && pdfOnlyByLecN.has(lec.sessionNumber)) {
			const pdf = pdfOnlyByLecN.get(lec.sessionNumber)!;
			// Merge: prefer the video's metadata, but attach the PDF URL
			lec.resolved = {
				pdfUrl: pdf.resolved.pdfUrl ?? lec.resolved.pdfUrl,
				transcriptUrl: pdf.resolved.transcriptUrl ?? lec.resolved.transcriptUrl,
				thumbnailUrl: lec.resolved.thumbnailUrl,
			};
			pdfOnlyByLecN.delete(lec.sessionNumber);
		}
	}
	// Any leftover PDFs without matching video → synthesize lecture entries
	for (const [n, { folder, r, resolved }] of pdfOnlyByLecN) {
		lectures.push({
			sessionNumber: n,
			sourceKey: r.title ?? folder,
			resource: {
				title: r.title ?? `Lecture ${n}`,
				resource_type: "Document",
				file_type: "application/pdf",
			},
			resolved,
			folderName: folder,
		});
	}

	// Sort lectures by sessionNumber; nulls to end
	lectures.sort((a, b) => {
		if (a.sessionNumber === null) return 1;
		if (b.sessionNumber === null) return -1;
		return a.sessionNumber - b.sessionNumber;
	});

	return {
		course,
		sessions,
		lectures,
		assignments,
		otherItems,
		assignmentsPageContent: assignmentsData?.content ?? null,
		warnings,
		coursePath,
		courseRelPath,
		folderName,
	};
}

// ---------- duration heuristics ----------
export function defaultLectureMinutes(r: ResourceData): number {
	if (r.youtube_key || r.archive_url) return 45;
	if (r.file_type === "application/pdf") return 30;
	return 30;
}

export function defaultAssignmentMinutes(_r: ResourceData): number {
	return 60;
}

File: src-tauri/src/importer.rs
use std::path::{Path, PathBuf};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use regex::Regex;
use std::sync::OnceLock;

use crate::error::{Error, Result};
use crate::models::{slugify, Instructor};

// ---------- parsed types ----------

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParsedSession {
    pub session_number: Option<String>,
    pub title: String,
    pub assignment_markers: Vec<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Resolved {
    pub pdf_url: Option<String>,        // on-disk path
    pub transcript_url: Option<String>, // on-disk path
    pub thumbnail_url: Option<String>,  // on-disk path or http
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResourceData {
    pub title: Option<String>,
    pub description: Option<String>,
    pub file: Option<String>,
    pub learning_resource_types: Vec<String>,
    pub resource_type: Option<String>,
    pub file_type: Option<String>,
    pub youtube_key: Option<String>,
    pub captions_file: Option<String>,
    pub transcript_file: Option<String>,
    pub thumbnail_file: Option<String>,
    pub archive_url: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParsedLecture {
    pub session_number: Option<i64>,
    pub source_key: String,
    pub resource: ResourceData,
    pub resolved: Resolved,
    pub folder_name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParsedAssignment {
    pub hw_number: Option<i64>,
    pub source_key: String,
    pub resource: ResourceData,
    pub resolved: Resolved,
    pub folder_name: String,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParsedCourse {
    pub course: ParsedCourseHeader,
    pub sessions: Vec<ParsedSession>,
    pub lectures: Vec<ParsedLecture>,
    pub assignments: Vec<ParsedAssignment>,
    pub other_items: Vec<ParsedOtherItem>,
    pub assignments_page_content: Option<String>,
    pub warnings: Vec<String>,
    pub course_path: String,
    pub course_rel_path: String,
    pub folder_name: String,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParsedCourseHeader {
    pub slug: String,
    pub title: String,
    pub description: Option<String>,
    pub course_number: Option<String>,
    pub extra_course_numbers: Option<String>,
    pub term: Option<String>,
    pub year: Option<String>,
    pub level: Option<String>,
    pub department_numbers: Vec<String>,
    pub instructors: Vec<Instructor>,
    pub topics: Vec<Vec<String>>,
    pub image_url: Option<String>,
    pub source_path: String,
    pub source_uid: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ParsedOtherItem {
    pub source_key: String,
    pub resource: ResourceData,
    pub resolved: Resolved,
    pub folder_name: String,
    #[serde(rename = "type")]
    pub kind: String,
}

// ---------- regex setup ----------

fn assignment_re() -> &'static Regex {
    static R: OnceLock<Regex> = OnceLock::new();
    R.get_or_init(|| Regex::new(r"(?i)Assignment\s+(\d+|\w+)\s+(out|due)").unwrap())
}

// ---------- helpers ----------

fn read_json(path: &Path) -> Option<Value> {
    let s = std::fs::read_to_string(path).ok()?;
    serde_json::from_str(&s).ok()
}

fn s(v: &Value, k: &str) -> Option<String> {
    v.get(k).and_then(|x| x.as_str()).map(|s| s.to_string())
}

fn arr_str(v: &Value, k: &str) -> Vec<String> {
    v.get(k)
        .and_then(|x| x.as_array())
        .map(|a| a.iter().filter_map(|x| x.as_str().map(String::from)).collect())
        .unwrap_or_default()
}

fn list_subdirs(dir: &Path) -> Vec<PathBuf> {
    match std::fs::read_dir(dir) {
        Ok(rd) => rd.filter_map(|e| e.ok())
            .filter(|e| e.file_type().map(|t| t.is_dir()).unwrap_or(false))
            .map(|e| e.path())
            .filter(|p| p.file_name().and_then(|n| n.to_str()).map(|n| !n.starts_with('.')).unwrap_or(true))
            .collect(),
        Err(_) => Vec::new(),
    }
}

fn session_number_from_folder(name: &str) -> Option<i64> {
    let l = name.to_lowercase();
    if let Some(caps) = regex::Regex::new(r"(?i)^lecture[-_ ]?(\d+)").unwrap().captures(&l) {
        return caps.get(1).and_then(|m| m.as_str().parse().ok());
    }
    if let Some(caps) = regex::Regex::new(r"(?i)[_-]?lec(\d+)").unwrap().captures(&l) {
        return caps.get(1).and_then(|m| m.as_str().parse().ok());
    }
    None
}

fn hw_number_from_folder(name: &str) -> Option<i64> {
    regex::Regex::new(r"(?i)[_-]?hw\s*(\d+)")
        .unwrap()
        .captures(name)
        .and_then(|c| c.get(1).and_then(|m| m.as_str().parse().ok()))
}

fn basename_of(file_url: &str) -> Option<String> {
    let clean = file_url.split(['?', '#']).next().unwrap_or("");
    let idx = clean.rfind('/').map(|i| i + 1).unwrap_or(0);
    let base = &clean[idx..];
    if base.is_empty() { None } else { Some(base.to_string()) }
}

fn resolve_static(course_path: &Path, file_url: Option<&str>) -> Option<String> {
    let base = basename_of(file_url?)?;
    let abs = course_path.join("static_resources").join(&base);
    if abs.is_file() {
        Some(abs.to_string_lossy().into_owned())
    } else {
        None
    }
}

fn resolve_resource(course_path: &Path, r: &ResourceData) -> Resolved {
    let pdf_url = resolve_static(course_path, r.file.as_deref());
    let transcript_url = resolve_static(course_path, r.transcript_file.as_deref());
    let thumbnail_url = if let Some(t) = &r.thumbnail_file {
        if t.starts_with("http://") || t.starts_with("https://") {
            Some(t.clone())
        } else {
            resolve_static(course_path, Some(t))
        }
    } else {
        None
    };
    Resolved { pdf_url, transcript_url, thumbnail_url }
}

// ---------- calendar parsing ----------

pub fn parse_calendar_content(raw: &str) -> Vec<ParsedSession> {
    if raw.is_empty() { return Vec::new(); }
    let header_re = regex::Regex::new(r"(?i)^(ses|#|session|topics|key\s+dates)").unwrap();
    let line_re = regex::Regex::new(
        r"^([A-Za-z]?\s*\d+\s*[-–~]\s*\d+|[A-Za-z]?\s*\d+)\s+(.*)$",
    ).unwrap();
    let assign_re = assignment_re();

    let mut lines: Vec<&str> = raw
        .lines()
        .map(|l| l.trim())
        .filter(|l| !l.is_empty())
        .collect();
    while lines.first().map(|first| header_re.is_match(first)).unwrap_or(false) {
        lines.remove(0);
    }

    let mut out = Vec::new();
    for line in lines {
        let caps = match line_re.captures(line) {
            Some(c) => c,
            None => continue,
        };
        let session_number = caps.get(1).map(|m| m.as_str().split_whitespace().collect::<String>());
        let rest = caps.get(2).map(|m| m.as_str()).unwrap_or("");

        let mut markers = Vec::new();
        for m in assign_re.captures_iter(rest) {
            let n = m.get(1).map(|x| x.as_str()).unwrap_or("");
            let v = m.get(2).map(|x| x.as_str()).unwrap_or("");
            markers.push(format!("{} {}", n.to_lowercase(), v.to_lowercase()));
        }

        let title: String = assign_re.replace_all(rest, "")
            .split_whitespace()
            .collect::<Vec<_>>()
            .join(" ");
        out.push(ParsedSession {
            session_number,
            title,
            assignment_markers: markers,
        });
    }
    out
}

// ---------- course discovery ----------

pub fn find_course_dirs(root: &Path) -> Vec<PathBuf> {
    let mut found = Vec::new();
    for sub in list_subdirs(root) {
        if sub.join("data.json").is_file() {
            found.push(sub);
            continue;
        }
        for nested in list_subdirs(&sub) {
            if nested.join("data.json").is_file() {
                found.push(nested);
            }
        }
    }
    found
}

// ---------- main parse ----------

pub fn parse_course(course_path: &Path, root: &Path) -> Result<ParsedCourse> {
    let warnings = Vec::new();
    let data_path = course_path.join("data.json");
    let data = read_json(&data_path).ok_or_else(|| Error::NotFound(format!("data.json at {}", course_path.display())))?;
    let folder_name = course_path.file_name().and_then(|n| n.to_str()).unwrap_or("").to_string();
    let course_rel_path = path_relative_to(root, course_path);

    let primary_num = s(&data, "primary_course_number");
    let term = s(&data, "term");
    let year = s(&data, "year");
    let slug_key = match (&primary_num, &term, &year) {
        (Some(p), Some(t), Some(y)) => format!("{}-{}-{}", p, t, y),
        (Some(p), _, _) => p.clone(),
        _ => folder_name.clone(),
    };
    let slug = slugify(&slug_key);

    // Image
    let image_url = s(&data, "image_src").and_then(|src| {
        if src.starts_with("http://") || src.starts_with("https://") {
            Some(src)
        } else {
            let rel = src.trim_start_matches("./").trim_start_matches('/');
            let abs = course_path.join(rel);
            if abs.is_file() {
                Some(abs.to_string_lossy().into_owned())
            } else {
                None
            }
        }
    });

    let instructors: Vec<Instructor> = data.get("instructors")
        .and_then(|v| v.as_array())
        .map(|a| a.iter().filter_map(|x| {
            if x.is_null() { return None; }
            Some(Instructor {
                first_name: x.get("first_name").and_then(|v| v.as_str()).map(String::from),
                last_name: x.get("last_name").and_then(|v| v.as_str()).map(String::from),
                middle_initial: x.get("middle_initial").and_then(|v| v.as_str()).map(String::from),
                salutation: x.get("salutation").and_then(|v| v.as_str()).map(String::from),
                title: x.get("title").and_then(|v| v.as_str()).map(String::from),
            })
        }).collect())
        .unwrap_or_default();

    let topics: Vec<Vec<String>> = data.get("topics")
        .and_then(|v| v.as_array())
        .map(|a| a.iter().filter_map(|x| x.as_array().map(|inner| {
            inner.iter().filter_map(|y| y.as_str().map(String::from)).collect()
        })).collect())
        .unwrap_or_default();

    let level = data.get("level").and_then(|v| v.as_array())
        .and_then(|a| a.first()).and_then(|v| v.as_str()).map(String::from);

    let course = ParsedCourseHeader {
        slug,
        title: s(&data, "course_title").unwrap_or_else(|| folder_name.clone()),
        description: s(&data, "course_description"),
        course_number: primary_num,
        extra_course_numbers: s(&data, "extra_course_numbers"),
        term,
        year,
        level,
        department_numbers: arr_str(&data, "department_numbers"),
        instructors,
        topics,
        image_url,
        source_path: course_rel_path.clone(),
        source_uid: s(&data, "site_uid"),
    };

    // Calendar
    let cal_data = read_json(&course_path.join("pages").join("calendar").join("data.json"));
    let sessions = cal_data.as_ref()
        .and_then(|v| v.get("content"))
        .and_then(|v| v.as_str())
        .map(parse_calendar_content)
        .unwrap_or_default();
    let assignments_data = read_json(&course_path.join("pages").join("assignments").join("data.json"));
    let assignments_page_content = assignments_data.as_ref()
        .and_then(|v| v.get("content"))
        .and_then(|v| v.as_str())
        .map(String::from);

    // Resources
    let resources_dir = course_path.join("resources");
    let resource_folders = list_subdirs(&resources_dir);

    let mut lectures: Vec<ParsedLecture> = Vec::new();
    let mut assignments: Vec<ParsedAssignment> = Vec::new();
    let mut other_items: Vec<ParsedOtherItem> = Vec::new();

    // First pass — collect lec-numbered PDF-only
    let mut pdf_only_by_lec: std::collections::HashMap<i64, (String, ResourceData, Resolved)> = std::collections::HashMap::new();

    for folder in &resource_folders {
        let fp = folder.join("data.json");
        let r_val = match read_json(&fp) { Some(v) => v, None => continue };
        let r = resource_from_value(&r_val);
        let resolved = resolve_resource(course_path, &r);
        let folder_name_local = folder.file_name().and_then(|n| n.to_str()).unwrap_or("").to_string();
        let lec_n = session_number_from_folder(&folder_name_local);
        let hw_n = hw_number_from_folder(&folder_name_local);
        let is_pdf = r.file_type.as_deref() == Some("application/pdf")
            || r.resource_type.as_deref() == Some("Document");

        if let Some(n) = lec_n {
            if is_pdf {
                pdf_only_by_lec.insert(n, (folder_name_local.clone(), r.clone(), resolved.clone()));
                continue;
            }
            lectures.push(ParsedLecture {
                session_number: Some(n),
                source_key: folder_name_local.clone(),
                resource: r.clone(),
                resolved: resolved.clone(),
                folder_name: folder_name_local,
            });
            continue;
        }
        if r.youtube_key.is_some() {
            lectures.push(ParsedLecture {
                session_number: None,
                source_key: folder_name_local.clone(),
                resource: r.clone(),
                resolved: resolved.clone(),
                folder_name: folder_name_local,
            });
            continue;
        }
        if let Some(n) = hw_n {
            assignments.push(ParsedAssignment {
                hw_number: Some(n),
                source_key: folder_name_local.clone(),
                resource: r.clone(),
                resolved: resolved.clone(),
                folder_name: folder_name_local,
            });
            continue;
        }
        if is_pdf && hw_n.is_none() {
            // pdf without lec/hw marker — treat as lecture-ish
            other_items.push(ParsedOtherItem {
                source_key: folder_name_local.clone(),
                resource: r.clone(),
                resolved: resolved.clone(),
                folder_name: folder_name_local,
                kind: r.resource_type.clone().unwrap_or_else(|| "other".into()),
            });
            continue;
        }
        other_items.push(ParsedOtherItem {
            source_key: folder_name_local.clone(),
            resource: r.clone(),
            resolved: resolved.clone(),
            folder_name: folder_name_local,
            kind: r.resource_type.clone().unwrap_or_else(|| "other".into()),
        });
    }

    // Merge pdf-only into video lectures
    let mut to_synthesize: Vec<(i64, String, ResourceData, Resolved)> = Vec::new();
    for lec in lectures.iter_mut() {
        if let Some(n) = lec.session_number {
            if let Some((_, pdf_r, pdf_resolved)) = pdf_only_by_lec.remove(&n) {
                lec.resolved = Resolved {
                    pdf_url: pdf_resolved.pdf_url.or(lec.resolved.pdf_url.clone()),
                    transcript_url: pdf_resolved.transcript_url.or(lec.resolved.transcript_url.clone()),
                    thumbnail_url: lec.resolved.thumbnail_url.clone(),
                };
                let _ = pdf_r;
            }
        }
    }
    for (n, (folder, r, resolved)) in pdf_only_by_lec.into_iter() {
        to_synthesize.push((n, folder, r, resolved));
    }
    for (n, folder, _r, resolved) in to_synthesize {
        lectures.push(ParsedLecture {
            session_number: Some(n),
            source_key: folder.clone(),
            resource: ResourceData {
                title: Some(format!("Lecture {}", n)),
                resource_type: Some("Document".into()),
                file_type: Some("application/pdf".into()),
                ..Default::default()
            },
            resolved,
            folder_name: folder,
        });
    }

    // Sort lectures by session_number (nulls last)
    lectures.sort_by(|a, b| match (a.session_number, b.session_number) {
        (Some(x), Some(y)) => x.cmp(&y),
        (Some(_), None) => std::cmp::Ordering::Less,
        (None, Some(_)) => std::cmp::Ordering::Greater,
        (None, None) => std::cmp::Ordering::Equal,
    });

    Ok(ParsedCourse {
        course,
        sessions,
        lectures,
        assignments,
        other_items,
        assignments_page_content,
        warnings,
        course_path: course_path.to_string_lossy().into_owned(),
        course_rel_path,
        folder_name,
    })
}

fn resource_from_value(v: &Value) -> ResourceData {
    ResourceData {
        title: s(v, "title"),
        description: s(v, "description"),
        file: s(v, "file"),
        learning_resource_types: arr_str(v, "learning_resource_types"),
        resource_type: s(v, "resource_type"),
        file_type: s(v, "file_type"),
        youtube_key: s(v, "youtube_key"),
        captions_file: s(v, "captions_file"),
        transcript_file: s(v, "transcript_file"),
        thumbnail_file: s(v, "thumbnail_file"),
        archive_url: s(v, "archive_url"),
    }
}

fn path_relative_to(base: &Path, p: &Path) -> String {
    let b = base.to_string_lossy().into_owned();
    let s = p.to_string_lossy().into_owned();
    if let Some(rest) = s.strip_prefix(&b) {
        rest.trim_start_matches('/').to_string()
    } else {
        s
    }
}

pub fn default_lecture_minutes(r: &ResourceData) -> i64 {
    if r.youtube_key.is_some() || r.archive_url.is_some() { 45 }
    else if r.file_type.as_deref() == Some("application/pdf") { 30 }
    else { 30 }
}

pub fn default_assignment_minutes(_r: &ResourceData) -> i64 { 60 }

// ---------- import runner ----------

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ImportResult {
    pub slug: String,
    pub title: String,
    pub updated: bool,
    pub ok: bool,
    pub error: Option<String>,
}

pub fn import_course_into_db(
    conn: &rusqlite::Connection,
    parsed: &ParsedCourse,
) -> Result<(String, bool)> {
    crate::import_runner::run(conn, parsed)
}

pub fn import_all(conn: &rusqlite::Connection, root: &Path) -> Result<Vec<ImportResult>> {
    let dirs = find_course_dirs(root);
    let mut results = Vec::new();
    for dir in dirs {
        let parsed = match parse_course(&dir, root) {
            Ok(p) => p,
            Err(e) => {
                results.push(ImportResult {
                    slug: dir.to_string_lossy().into_owned(),
                    title: dir.to_string_lossy().into_owned(),
                    updated: false,
                    ok: false,
                    error: Some(e.to_string()),
                });
                continue;
            }
        };
        let slug = parsed.course.slug.clone();
        let title = parsed.course.title.clone();
        match import_course_into_db(&conn, &parsed) {
            Ok((_id, updated)) => results.push(ImportResult { slug, title, updated, ok: true, error: None }),
            Err(e) => results.push(ImportResult { slug, title, updated: false, ok: false, error: Some(e.to_string()) }),
        }
    }
    Ok(results)
}


